types/types.cc

```diff
@@ -1834,6 +1837,7 @@ struct serialize_visitor {
     bytes::iterator& out;
     ;
     void operator()(const reversed_type_impl& t, const void* v) { return serialize(*t.underlying_type(), v, out); }
+    void operator()(const counter_type_impl& t, const void* v) { return serialize(*long_type, v, out); }
```

Comment:
Usually CQL classes for type impls have a member called `native_type`.
I think this specifies the native representation that we use after deserializing, the `void*` is probably a pointer to a value of this type.

`long_type_impl::native_type` is `maybe_empty<int64_t>`.
I think we should add one for the counter type as well, so that the situation is clear:
```diff
diff --git a/concrete_types.hh b/concrete_types.hh
index 5dc98a322..93a78feea 100644
--- a/concrete_types.hh
+++ b/concrete_types.hh
@@ -25,6 +25,8 @@ struct empty_type_impl final : public abstract_type {
 
 struct counter_type_impl final : public abstract_type {
     counter_type_impl();
+
+    using native_type = maybe_empty<int64_t>;
 };
```

Reply:
Maybe, although, if it worked without this "native_type" alias, why bother?
Note that I deliberately used "long_type" in this patch and didn't try to open-code long-type's implementation, so that I not make any mistakes. Setting "native_type" and using it is yet place where things can become inconsistent (what if I set native_type and it's not the same as long_type's native_type)?
But I'll try this out and see how it works out.